Take steps to reduce database queries

geoversed opened the pull request:
Proposes the following:
- [x] Allow the `transform` func caller to specify an arbitrary number of fields they wish to receive alongside just the name.
- [x] ~~The same as the above but for item partial matching. For instance, we retrieved more fields about the item such as whether it is available in the shop at once.~~
- [x] Remove the ability for users to provide the new content to edit the tag with OR find a better route of dealing with the situation. Removing the parameter is a big cost to UX especially if it is only a short amount of content to update to.
- [x] #168

geoversed committed:
Take steps to reduce database queries
Proposes the following:
- [ ] Allow the `transform` func caller to specify an arbitrary number of fields they wish to receive alongside just the name.
  - Do not rely on discord by storing the metadata for tags in the resulting options themselves. An alternative suggestion is to store a dictionary like we have with kona searching for bookmarks.
- [ ] Remove the ability for users to provide the new content to edit the tag with OR find a better route of dealing with the situation. Removing the parameter is a big cost to UX especially if it is only a short amount of content to update to.
  - It would also be nice if something similar happened to item input transformers. For instance, we retrieved more fields about the item such as whether it is available in the shop at once.
- [ ] Deal with #168

SGA-A commented:
In general, you want to have functions that return the same type and the same thing every time.

Dynamically changing the return type based on the inputs makes debugging more difficult from my experience. If we want to implement these features, it needs to be thoroughly battle tested. Many days of writing the code, and many weeks of testing and iterating over our code design choices and mistakes.

I cannot recall seeing instances of functions by others where the return type and input types are both dynamic in anyone else's code, though in some functions of our code we already have this practice in place (in particular the SQL table query updating functions).

SGA-A commented:
> In general, you want to have functions that return the same type and the same thing every time.
> 
> Dynamically changing the return type based on the inputs makes debugging more difficult from my experience. If we want to implement these features, it needs to be thoroughly battle tested. Many days of writing the code, and many weeks of testing and iterating over our code design choices and mistakes.
> 
> I cannot recall seeing instances of functions by others where the return type and input types are both dynamic in anyone else's code, though in some functions of our code we already have this practice in place (in particular the SQL table query updating functions).

So in reality the return types are not actually dynamic. The `transform` function instead returns a tuple containing `sqlite3.Row` and the interaction instead of just the tag name (`str`) and interaction. The columns that will appear are the only things which change but the return type is always the same. This approach is good for now as you can access a field value easily by using the index (provided you know the location of the field you want to get which is faster) or the name of the column when the query is made. Former is simplistic, latter is more readable. We will use both depending on the circumstances.

geoversed committed:
Reduce database queries in tags + more

Alterations to the transform and partial matching functions made to reduce the need for more queries in sequential lines of code.

In this commit, make dev tools command responses public and add a new level up prompt.

SGA-A commented:
The PR is not yet ready but @serephenna can you review the current changes.

I most likely will not be able to continue with the other pending tasks until the following Friday.

geoversed committed:
Fix MatchView default select options not being set

SGA-A commented:
Second task is no longer planned, causes problems for specific commands like the item command where it needs access to many details about the item across multiple tables.

geoversed committed:
Item transformer improvements

Caching has not been implemented, simple tweaks were only made.

geoversed committed:
Unsubscribe from message intents

Changes to commands were made to accomodate this update.

Commit:
Change credits hyperlink

geoversed committed:
Config initialization changes

Configuration is now loaded within the main function instead of the bot constructor. Owner IDs are also no longer hardcoded and is taken via the cached API call.

geoversed committed:
Remove Jeyy API commands

The API is too unstable for public use

Commit:
Update CODEOWNERS

Point code owner to new renamed username

geoversed commented on src/tags.py:
```diff
@@ -393,40 +349,34 @@ async def remove_id(itx: Interaction, tag_id: int) -> None:
+    # Unpacking the row
+    embed.title, rowid, uses, owner_id, time_created = row
 
     user = (
         itx.client.get_user(owner_id)
         or (await itx.client.fetch_user(owner_id))
     )
```
Suggestion: make a get or fetch method within the subclassed bot and use that.

It would look much cleaner and can be reused in the leaderboards too when that arrives

geoversed committed:
Remove bot variable from Jeyy API

geoversed committed:
Apply suggestion

geoversed committed:
Run bot using winloop event policy

Experimenting with alternatives to asyncio event loop as #171 mentions.

geoversed committed:
Update requirements to include winloop

geoversed committed:
Fix for expiry multiplier insertion attempts

Function assumes an expiry column exists within the database but this was removed recently.

geoversed committed:
Cooldown command optimizations

geoversed committed:
Store socketstats in database

geoversed committed:
Initial caching commit

Commit:
Switch to Linux

Closes #171

Commit:
Add LRU cache for items and tags

Commit:
Update function names

Commit:
Append newlines to all files

Done for consistency
